### DISCUSSION TITLE
Using autodiff with custom classes

### DISCUSSION THREAD

**Discussion by cpp977** [autodiff/autodiff#199] [Q&A] [answered]
Hey,

the autodiff library allows the direct use of Eigen types. 
Is there some documentation of what is needed if one wants to use it with a different custom type of the form `template<typename Scalar, typename OtherStuff> MyClass;`?

Thanks!
Matthias

**Comment by allanleal** [comment, marked as answer]
I'm not sure I understand what you want. Could you please elaborate more? A simple example perhaps.

**Reply by cpp977** [reply]
What you can do with autodiff now:
```
autodiff::var my_func(const Eigen::Matrix<autodiff::var, Eigen::Dynamic, 1>& input);
Eigen::Matrix<autodiff::var, Eigen::Dynamic, 1> vec (); vec.setRandom();
auto y = my_func(vec);
Eigen::Matrix<autodiff::var, Eigen::Dynamic, 1> grad = autodiff::gradient(y,vec);
```
I want to do exactly the same with `MyClass<autodiff::var, OtherStuff>`.
The question is if you have a documentation of how to achieve that (requirements on `MyClass` and traits or similar stuff for autodiff) or if you have documentation about how you did this for the Eigen type.

I hope that clarifies my question.

**Reply by allanleal** [reply]
I'm not sure what your `MyClass` does, but maybe this is irrelevant. If `MyClass` is supposed to act as a vector/array (like it is done with Eigen), then there is no specific docs at the moment; inspecting the source code of `autodiff` (in the `eigen.hpp` file) and trying to replicate the functionality there for your own type should do the trick (hopefully!).

**Reply by cpp977** [reply]
Yes its an array like class with algebraic operations defined (Scalar * array and also array * array). I will have a look to this header.
Thanks for the quick reply!
